### ISSUE TITLE
Tutorial 5: Error when executing 'asm("sti");'

### ISSUE THREAD

**Issue by jj01101010** [AlgorithMan-de/wyoos#14]
If i execute the line you see above. VirtualBox is crashing and spamming my terminal with many lines but i dont know 1. Why it is crashing and 2. How can I fix this crash

Pls help me i am lost

**Comment by jj01101010** [comment]
But i have to say your tutorial series is GREAT pls never leave this project

**Comment by sgermain06** [comment]
Hey, I have the same problem. If I comment out the sti call, it doesn't crash, but if I don't, as soon as it calls it, VirtualBox crashes. Were you ever able to fix this? Anyone can help?

**Comment by sgermain06** [comment]
Ah! I found out what was happening. In my gdt constructor, I had inverted the indexes of the array I sent to the `lgdt` command. After I switched those around, it stopped crashing. Now, I have another problem though. When I call `sti`, not only I don't receive any interrupt calls to my `HandleInterrupt` function, but after like 5-10 seconds, my screens clear and fills with some weird characters. I'm trying to compare my code with the code in this git repository and, so far, I seem to be matching. I'm sure I'm missing something stupid...

**Comment by sgermain06** [comment]
Yeah, I found what I did. In the interrupts.cpp, I set the handlerAddressLowBits twice instead of setting the handlerAddressLowBits and handlerAddressHighBits. Everything is working fine now.

**Comment by fluecs** [comment]
Same issue. Can't seem to fix it.

**Comment by M1sterPl0w** [comment]
@HayJayDee  @Ilc-Ilia-Developer  Is it possible to show your code? Maybe I can take a look.

**Comment by StjepanBM1** [comment]
Does anyone know how to fix this issue?

**Comment by ssagar-linux** [comment]
I resolved following issue via:-
-GlobalDescriptorTable::SegmentDescriptor::SegmentDescriptor(uint32_t limit, uint32_t base , uint8_t type)
+GlobalDescriptorTable::SegmentDescriptor::SegmentDescriptor(uint32_t base, uint32_t limit, uint8_t type)

mistakenly using base & limit interchangeably

**Comment by yutojama** [comment]
1. check out constructor of GlobalDescriptorTable(), make sure i[0] = sizeof(GlobalDescriptorTable) << 16; i[1] = (uint32_t) i[2];
2. check all structs with __attribute__((packed)), make sure everything is placed just in the order as tutorial and nothing is randomly swapped. For me, I changed sequence of size and base in InterruptDescriptorTablePointer and that led to error.

**Comment by Zuhaitz-dev** [comment]
As @yutojama said, change i[0] and i[1] in gdt.cpp. Then the rest is typos surely. In my case I had a stupid mistake, and instead of writing:

```cpp
GlobalDescriptorTable::SegmentDescriptor::SegmentDescriptor(uint32_t base, uint32_t limit, uint8_t flags)
{
    uint8_t* target = (uint8_t*)this;

    if (limit <= 65536)
    {
        target[6] = 0x40;
    }
    else
    {
        if ((limit & 0xFFF) != 0xFFF)
            limit = (limit >> 12) - 1;
        else
            limit = limit >> 12;

        target[6] = 0xC0;                        // CORRECT
    }

// ...
```

I wrote:

```cpp
GlobalDescriptorTable::SegmentDescriptor::SegmentDescriptor(uint32_t base, uint32_t limit, uint8_t flags)
  {
      uint8_t* target = (uint8_t*)this;
  
      if (limit <= 65536)
      {
          target[6] = 0x40;
      }
      else 
      {
          if ((limit & 0xFFF) != 0xFFF)
          {
              limit = (limit >> 12) - 1;
          }
          else
          {
              limit = limit >> 12;
  
              target[6] = 0xC0;                        // WRONG: not inside the second else
          }
      }
```

Besides that, I would check carefully `interruptstubs.s` and check the macros. Don't forget:

```bash
nm interrupts.o
```
